File: frontend/lib/data/local/tables/operations.dart
/// Housekeeping, maintenance et pieces jointes du terrain.
library;

import 'package:drift/drift.dart';

import '../columns.dart';
import '../enums.dart';

/// Tache de nettoyage (F2.1, F2.2).
///
/// Les horodatages de debut et de fin ne servent pas qu'au suivi : ils
/// donnent la duree reelle par type de tache, seul moyen de dimensionner une
/// equipe d'etage et d'alimenter les indicateurs de performance du module
/// Direction (F5.1).
@DataClassName('HousekeepingTaskRow')
class HousekeepingTasks extends Table with SyncedTableColumns, HotelScoped {
  TextColumn get roomId => text().withLength(min: 36, max: 36)();
  TextColumn get type => textEnum<HousekeepingTaskType>()();
  TextColumn get status =>
      textEnum<TaskStatus>().withDefault(const Constant('PENDING'))();
  TextColumn get priority =>
      textEnum<Priority>().withDefault(const Constant('NORMAL'))();
  TextColumn get businessDate => text().withLength(max: 10)();

  TextColumn get assignedTo => text().nullable()();
  DateTimeColumn get assignedAt => dateTime().nullable()();
  DateTimeColumn get startedAt => dateTime().nullable()();
  DateTimeColumn get finishedAt => dateTime().nullable()();
  IntColumn get durationMinutes => integer().nullable()();

  TextColumn get inspectedBy => text().nullable()();
  DateTimeColumn get inspectedAt => dateTime().nullable()();
  TextColumn get notes => text().nullable()();
}

/// Point de controle d'une fiche de tache, telle qu'imprimee en F2.5.
@DataClassName('HousekeepingTaskItemRow')
class HousekeepingTaskItems extends Table with SyncedTableColumns {
  TextColumn get taskId =>
      text().references(HousekeepingTasks, #id, onDelete: KeyAction.cascade)();
  TextColumn get label => text().withLength(max: 160)();
  BoolColumn get isDone => boolean().withDefault(const Constant(false))();
  TextColumn get remark => text().withLength(max: 255).nullable()();
  IntColumn get sortOrder => integer().withDefault(const Constant(0))();
}

/// Consommables et linge utilises sur une tache (F2.4).
///
/// Fait le pont entre le housekeeping et les stocks : chaque ligne donnera un
/// mouvement de sortie sur le magasin de l'etage.
@DataClassName('AmenityConsumptionRow')
class AmenityConsumptions extends Table with SyncedTableColumns {
  TextColumn get taskId =>
      text().references(HousekeepingTasks, #id, onDelete: KeyAction.cascade)();
  TextColumn get productId => text().withLength(min: 36, max: 36)();

  /// Quantite entiere. L'unite est portee par le produit (`products.unit`).
  IntColumn get quantity => integer().withDefault(const Constant(1))();
}

/// Equipement suivi : climatiseur, chauffe-eau, televiseur, ascenseur.
///
/// Permet l'historique par equipement demande en F4.4, qui ne se deduit pas de
/// l'historique par chambre : un climatiseur peut etre deplace d'une chambre a
/// une autre, et un ascenseur n'appartient a aucune chambre.
@DataClassName('EquipmentRow')
class Equipments extends Table
    with SyncedTableColumns, HotelScoped, RefTableColumns {
  TextColumn get code => text().withLength(max: 32)();
  TextColumn get label => text().withLength(max: 160)();
  TextColumn get category => text().withLength(max: 64).nullable()();
  TextColumn get roomId => text().nullable()();
  TextColumn get location => text().withLength(max: 120).nullable()();
  TextColumn get brand => text().withLength(max: 80).nullable()();
  TextColumn get model => text().withLength(max: 80).nullable()();
  TextColumn get serialNumber => text().withLength(max: 80).nullable()();
  TextColumn get installedAt => text().withLength(max: 10).nullable()();
  TextColumn get warrantyUntil => text().withLength(max: 10).nullable()();
  TextColumn get supplierId => text().nullable()();
}

/// Signalement et intervention (F2.3, F4.1 a F4.4).
///
/// `blocksRoom` fait le lien avec `Rooms.isOutOfOrder` : ouvrir un ticket
/// bloquant sort la chambre de la vente, le cloturer l'y remet. Sans ce lien
/// explicite, une chambre reste hors service parce que personne ne pense a
/// rebasculer le drapeau une fois la reparation faite -- et l'hotel perd des
/// nuitees sans s'en apercevoir.
@DataClassName('MaintenanceTicketRow')
class MaintenanceTickets extends Table with SyncedTableColumns, HotelScoped {
  TextColumn get number => text().withLength(max: 32)();
  TextColumn get roomId => text().nullable()();
  TextColumn get equipmentId => text().nullable()();
  TextColumn get location => text().withLength(max: 120).nullable()();
  TextColumn get category => text().withLength(max: 64).nullable()();

  TextColumn get title => text().withLength(max: 160)();
  TextColumn get description => text().nullable()();
  TextColumn get priority =>
      textEnum<Priority>().withDefault(const Constant('NORMAL'))();
  TextColumn get status =>
      textEnum<TicketStatus>().withDefault(const Constant('OPEN'))();

  TextColumn get reportedBy => text().nullable()();
  DateTimeColumn get reportedAt => dateTime().nullable()();
  TextColumn get assignedTo => text().nullable()();
  DateTimeColumn get assignedAt => dateTime().nullable()();
  DateTimeColumn get resolvedAt => dateTime().nullable()();
  DateTimeColumn get closedAt => dateTime().nullable()();
  TextColumn get resolution => text().nullable()();

  /// Cout en francs CFA, entiers.
  IntColumn get cost => integer().withDefault(const Constant(0))();
  BoolColumn get blocksRoom => boolean().withDefault(const Constant(false))();
}

/// Passage d'un technicien sur un ticket.
///
/// Un ticket demande souvent plusieurs passages : diagnostic, commande de
/// piece, reparation. Les tracer separement donne le temps reellement passe et
/// le cout des pieces, que le seul ticket ne permet pas de reconstituer.
@DataClassName('MaintenanceInterventionRow')
class MaintenanceInterventions extends Table with SyncedTableColumns {
  TextColumn get ticketId =>
      text().references(MaintenanceTickets, #id, onDelete: KeyAction.cascade)();
  TextColumn get technicianId => text().nullable()();
  DateTimeColumn get startedAt => dateTime().nullable()();
  DateTimeColumn get endedAt => dateTime().nullable()();
  TextColumn get description => text().nullable()();

  /// Pieces utilisees, en JSON.
  TextColumn get partsUsed => text().nullable()();
  IntColumn get cost => integer().withDefault(const Constant(0))();
}

/// Piece jointe generique : photo de panne, document, rapport.
///
/// Meme principe que les signatures : le fichier existe d'abord sur la
/// tablette, puis recoit une URL serveur. Le televersement est suivi
/// separement de la synchronisation des lignes metier, parce qu'une photo de
/// trois megaoctets prise dans un couloir sans reseau ne doit pas retarder la
/// remontee du ticket urgent auquel elle est attachee, qui ne pese lui que
/// quelques centaines d'octets.
@DataClassName('AttachmentRow')
class Attachments extends Table with SyncedTableColumns {
  TextColumn get entityTable => text().withLength(max: 64)();
  TextColumn get entityId => text().withLength(min: 36, max: 36)();
  TextColumn get kind =>
      text().withLength(max: 32).withDefault(const Constant('PHOTO'))();
  TextColumn get filePathLocal => text().withLength(max: 255).nullable()();
  TextColumn get fileUrl => text().withLength(max: 512).nullable()();
  TextColumn get mimeType => text().withLength(max: 80).nullable()();
  IntColumn get sizeBytes => integer().nullable()();
  TextColumn get uploadState =>
      textEnum<UploadState>().withDefault(const Constant('PENDING'))();
  TextColumn get caption => text().withLength(max: 255).nullable()();
  DateTimeColumn get capturedAt => dateTime().nullable()();
  TextColumn get capturedBy => text().nullable()();
}

File: backend/app/models/operations.py
"""Housekeeping et maintenance, plus les pieces jointes du terrain."""

from __future__ import annotations

import datetime as dt
import uuid

from sqlalchemy import (
    BigInteger,
    Boolean,
    Date,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import HotelScoped, RefBase, SyncBase
from app.models.enums import (
    HousekeepingTaskType,
    Priority,
    TaskStatus,
    TicketStatus,
    UploadState,
)
from app.models.rooms import _enum


class HousekeepingTask(SyncBase, HotelScoped):
    """Tache de nettoyage (F2.1, F2.2).

    Les horodatages de debut et de fin ne servent pas qu'au suivi : ils
    alimentent la duree moyenne par type de tache, seul moyen de dimensionner
    une equipe d'etage et d'alimenter les indicateurs de performance du
    module Direction (F5.1).
    """

    __tablename__ = "housekeeping_tasks"
    __table_args__ = (
        Index("ix_housekeeping_tasks_date_status", "business_date", "status"),
        Index("ix_housekeeping_tasks_assigned_to", "assigned_to"),
    )

    room_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("rooms.id", ondelete="CASCADE"), index=True
    )
    type: Mapped[HousekeepingTaskType] = mapped_column(
        _enum(HousekeepingTaskType, "housekeeping_task_type")
    )
    status: Mapped[TaskStatus] = mapped_column(
        _enum(TaskStatus, "task_status"), default=TaskStatus.PENDING
    )
    priority: Mapped[Priority] = mapped_column(
        _enum(Priority, "priority"), default=Priority.NORMAL
    )
    business_date: Mapped[dt.date] = mapped_column(Date)

    assigned_to: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), default=None
    )
    assigned_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    started_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    finished_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    duration_minutes: Mapped[int | None] = mapped_column(default=None)

    inspected_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), default=None
    )
    inspected_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    notes: Mapped[str | None] = mapped_column(Text, default=None)

    items: Mapped[list[HousekeepingTaskItem]] = relationship(
        back_populates="task", lazy="selectin"
    )


class HousekeepingTaskItem(SyncBase):
    """Point de controle d'une fiche de tache (imprimee, cf. F2.5)."""

    __tablename__ = "housekeeping_task_items"

    task_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("housekeeping_tasks.id", ondelete="CASCADE"), index=True
    )
    label: Mapped[str] = mapped_column(String(160))
    is_done: Mapped[bool] = mapped_column(Boolean, default=False)
    remark: Mapped[str | None] = mapped_column(String(255), default=None)
    sort_order: Mapped[int] = mapped_column(default=0)

    task: Mapped[HousekeepingTask] = relationship(back_populates="items")


class AmenityConsumption(SyncBase):
    """Consommables et linge utilises sur une tache (F2.4).

    Fait le pont entre le housekeeping et les stocks : chaque ligne genere un
    mouvement de sortie sur le magasin de l'etage.
    """

    __tablename__ = "amenity_consumptions"

    task_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("housekeeping_tasks.id", ondelete="CASCADE"), index=True
    )
    product_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("products.id", ondelete="RESTRICT"), index=True
    )
    quantity: Mapped[int] = mapped_column(Integer, default=1)


class Equipment(RefBase, HotelScoped):
    """Equipement suivi : climatiseur, chauffe-eau, televiseur, ascenseur.

    Permet l'historique par equipement demande en F4.4, qui ne se deduit pas
    de l'historique par chambre : un climatiseur peut etre deplace, et un
    ascenseur n'appartient a aucune chambre.
    """

    __tablename__ = "equipments"
    __table_args__ = (UniqueConstraint("hotel_id", "code"),)

    code: Mapped[str] = mapped_column(String(32))
    label: Mapped[str] = mapped_column(String(160))
    category: Mapped[str | None] = mapped_column(String(64), default=None)
    room_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("rooms.id", ondelete="SET NULL"), default=None, index=True
    )
    location: Mapped[str | None] = mapped_column(String(120), default=None)
    brand: Mapped[str | None] = mapped_column(String(80), default=None)
    model: Mapped[str | None] = mapped_column(String(80), default=None)
    serial_number: Mapped[str | None] = mapped_column(String(80), default=None)
    installed_at: Mapped[dt.date | None] = mapped_column(Date, default=None)
    warranty_until: Mapped[dt.date | None] = mapped_column(Date, default=None)
    supplier_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("suppliers.id", ondelete="SET NULL"), default=None
    )


class MaintenanceTicket(SyncBase, HotelScoped):
    """Signalement et intervention (F2.3, F4.1 a F4.4).

    `blocks_room` fait le lien avec `rooms.is_out_of_order` : ouvrir un ticket
    bloquant sort la chambre de la vente, le cloturer l'y remet. Sans ce lien
    explicite, une chambre reste hors service parce que personne ne pense a
    reactiver le drapeau apres reparation.
    """

    __tablename__ = "maintenance_tickets"
    __table_args__ = (
        UniqueConstraint("hotel_id", "number"),
        Index("ix_maintenance_tickets_status_priority", "status", "priority"),
        Index("ix_maintenance_tickets_assigned_to", "assigned_to"),
    )

    number: Mapped[str] = mapped_column(String(32))
    room_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("rooms.id", ondelete="SET NULL"), default=None, index=True
    )
    equipment_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("equipments.id", ondelete="SET NULL"), default=None, index=True
    )
    location: Mapped[str | None] = mapped_column(String(120), default=None)
    category: Mapped[str | None] = mapped_column(String(64), default=None)

    title: Mapped[str] = mapped_column(String(160))
    description: Mapped[str | None] = mapped_column(Text, default=None)
    priority: Mapped[Priority] = mapped_column(
        _enum(Priority, "priority"), default=Priority.NORMAL
    )
    status: Mapped[TicketStatus] = mapped_column(
        _enum(TicketStatus, "ticket_status"), default=TicketStatus.OPEN
    )

    reported_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), default=None
    )
    reported_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    assigned_to: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), default=None
    )
    assigned_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    resolved_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    closed_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    resolution: Mapped[str | None] = mapped_column(Text, default=None)
    cost: Mapped[int] = mapped_column(BigInteger, default=0)
    blocks_room: Mapped[bool] = mapped_column(Boolean, default=False)

    interventions: Mapped[list[MaintenanceIntervention]] = relationship(
        back_populates="ticket", lazy="selectin"
    )


class MaintenanceIntervention(SyncBase):
    """Passage d'un technicien sur un ticket.

    Un ticket peut demander plusieurs passages (diagnostic, commande de piece,
    reparation). Les tracer separement donne le temps reellement passe et le
    cout des pieces, que le seul ticket ne permet pas de reconstituer.
    """

    __tablename__ = "maintenance_interventions"

    ticket_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("maintenance_tickets.id", ondelete="CASCADE"), index=True
    )
    technician_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), default=None
    )
    started_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    ended_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    description: Mapped[str | None] = mapped_column(Text, default=None)
    parts_used: Mapped[dict | None] = mapped_column(JSONB, default=None)
    cost: Mapped[int] = mapped_column(BigInteger, default=0)

    ticket: Mapped[MaintenanceTicket] = relationship(back_populates="interventions")


class Attachment(SyncBase):
    """Piece jointe generique : photo de panne, document, rapport.

    Meme principe que `signatures` : le fichier existe d'abord sur la tablette,
    puis recoit une URL serveur. `upload_state` est suivi par un worker distinct
    de la synchronisation des lignes metier, parce qu'une photo de 3 Mo prise
    dans un couloir sans reseau ne doit pas bloquer la remontee d'un ticket
    urgent qui, lui, ne pese que quelques centaines d'octets.
    """

    __tablename__ = "attachments"
    __table_args__ = (
        Index("ix_attachments_entity", "entity_table", "entity_id"),
        Index("ix_attachments_upload_state", "upload_state"),
    )

    entity_table: Mapped[str] = mapped_column(String(64))
    entity_id: Mapped[uuid.UUID] = mapped_column()
    kind: Mapped[str] = mapped_column(String(32), default="PHOTO")
    file_path_local: Mapped[str | None] = mapped_column(String(255), default=None)
    file_url: Mapped[str | None] = mapped_column(String(512), default=None)
    mime_type: Mapped[str | None] = mapped_column(String(80), default=None)
    size_bytes: Mapped[int | None] = mapped_column(default=None)
    upload_state: Mapped[UploadState] = mapped_column(
        _enum(UploadState, "upload_state"), default=UploadState.PENDING
    )
    caption: Mapped[str | None] = mapped_column(String(255), default=None)
    captured_at: Mapped[dt.datetime | None] = mapped_column(default=None)
    captured_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), default=None
    )
